### PR TITLE
Spirv-fuzz: Add tests for MaybeGet* functionalities in fuzzer util

### PR DESCRIPTION
**Pull request by sliu-UIUC**
This PR will solve #4179 by adding unit testing for all maybeGet* functions in fuzzer_util. 

Currently several tests remain in question. Details are in the comments.

### CODE REVIEW

File: source/fuzz/fuzzer_util.cpp [hunk 1]
```diff
@@ -1130,9 +1130,9 @@ uint32_t MaybeGetVectorType(opt::IRContext* ir_context,
   assert(component_type &&
          (component_type->AsInteger() || component_type->AsFloat() ||
           component_type->AsBool()) &&
-         "|component_type_id| is invalid");
+         "'component_type_id' is invalid");
   assert(element_count >= 2 && element_count <= 4 &&
-         "Precondition: component count must be in range [2, 4].");
+         "Precondition: component count must be in range 2 to 4.");
```
Comment (afd) [inline source/fuzz/fuzzer_util.cpp:1135]: Please revert these changes. You can adapt the death test so that it can recognise these strings.

File: test/fuzz/fuzzerutil_test.cpp [hunk 2]
```diff
@@ -0,0 +1,1519 @@
...
+  // The vector type with |element_count| 3 and |component_type_id| 7
+  // is not present in the module.
+  ASSERT_EQ(0, fuzzerutil::MaybeGetVectorType(ir_context, 7, 3));
+
+#ifndef NDEBUG
+  // It should abort with |component_type_id| of 100
+  // |component_type_id| must be a valid result id of an OpTypeInt,
+  // OpTypeFloat or OpTypeBool instruction in the module.
+  ASSERT_DEATH(fuzzerutil::MaybeGetVectorType(ir_context, 100, 4),
+               "'component_type_id' is invalid");
```
Comment (afd) [inline test/fuzz/fuzzerutil_test.cpp:1281]: Try:

```
  ASSERT_DEATH(fuzzerutil::MaybeGetVectorType(ir_context, 100, 4),
               "\\|component_type_id\\| is invalid");
```

File: test/fuzz/fuzzerutil_test.cpp [hunk 3]
```diff
@@ -0,0 +1,1519 @@
...
+
+  // It should abort with |element_count| of 5.
+  // |element_count| must be in the range [2,4].
+  ASSERT_DEATH(fuzzerutil::MaybeGetVectorType(ir_context, 7, 5),
+               "Precondition: component count must be in range 2 to 4.");
```
Comment (afd) [inline test/fuzz/fuzzerutil_test.cpp:1286]: Try:

```
  ASSERT_DEATH(fuzzerutil::MaybeGetVectorType(ir_context, 7, 5),
               "Precondition: component count must be in range \\[2, 4\\].");
```

File: test/fuzz/fuzzerutil_test.cpp (hunk 2 above)
Comment (afd) [reply test/fuzz/fuzzerutil_test.cpp:1281]: i.e., use "\|" instead of "|", to avoid the special meaning of "|". But because "\" has a special meaning in a C++ string we need to escape it in the string literal to get a plain "\", hence the use of "\\".

File: test/fuzz/fuzzerutil_test.cpp [hunk 4]
```diff
@@ -0,0 +1,1519 @@
...
+  uint32_t maybe_bool_id = fuzzerutil::MaybeGetZeroConstant(
+      ir_context, transformation_context, 6, false);
+  // The id of a 32 bit float constant will be returned given the float type
+  // id 14.
+  uint32_t maybe_float_id = fuzzerutil::MaybeGetZeroConstant(
+      ir_context, transformation_context, 14, false);
+  uint32_t maybe_signed_int_id = fuzzerutil::MaybeGetZeroConstant(
+      ir_context, transformation_context, 20, false);
+  uint32_t maybe_unsigned_int_id = fuzzerutil::MaybeGetZeroConstant(
+      ir_context, transformation_context, 26, false);
```
Comment (afd) [inline test/fuzz/fuzzerutil_test.cpp:1496]: Note that there is no zero constant with type 26. The instruction with id 29 is a *constant* with type 26, but it is not a *zero* constant: it has the value 100.

Perhaps there are similar problems with the other assertions - I didn't check.

Please take a careful look and adapt accordingly.

**Review by afd — CHANGES_REQUESTED** [review]
Thanks - this looks like a good start!

I've added some comments on some of the problems.

File: source/fuzz/fuzzer_util.cpp (hunk 1 above)
Comment (sliu-UIUC) [reply source/fuzz/fuzzer_util.cpp:1135]: Reverted

File: test/fuzz/fuzzerutil_test.cpp (hunk 3 above)
Comment (sliu-UIUC) [reply test/fuzz/fuzzerutil_test.cpp:1286]: I remembered doing something similar previously. I just don't know why the same line of assertion would work in ``fuzzerutil`` but not in my test code, i.e., I have to put \\ in front of special symbols in my test. 
Maybe assert and ASSERT are compiled and evaluated differently?

File: test/fuzz/fuzzerutil_test.cpp [hunk 5]
```diff
@@ -0,0 +1,1519 @@
...
+  spvtools::ValidatorOptions validator_options;
+  ASSERT_TRUE(fuzzerutil::IsValidAndWellFormed(context.get(), validator_options,
+                                               kConsoleMessageConsumer));
+  TransformationContext transformation_context(
+      MakeUnique<FactManager>(context.get()), validator_options);
+
+  opt::IRContext* ir_context = context.get();
+  // A 32 bit float constant with width 14 exists and the id is 17.
+  // ASSERT_EQ(17, fuzzerutil::MaybeGetFloatConstant(ir_context,
+  // transformation_context, std::vector<uint32_t> {123}, 32, false));
```
[the remark this answers is not in the document]
Comment (sliu-UIUC) [reply test/fuzz/fuzzerutil_test.cpp:476]: Actually those are the questions that I have regarding what word vector would be for floats and negative constants. They were the remaining questions I had in the previous PR. 
I assume word vector would entail information about the constant's value. Thus words vector for positive single constant would be that number exactly, but then I'm not sure how it would represent float and negative. Floats have different number of decimal places and negative constants need a number to indicate that its negative.

File: test/fuzz/fuzzerutil_test.cpp (hunk 4 above)
Comment (sliu-UIUC) [reply test/fuzz/fuzzerutil_test.cpp:1496]: Thanks. I think I mistook the zero constant for the signed one with the unsigned one and had the error. Now its fixed.
However, here is the float part of the shader script: 
```
         %14 = OpTypeFloat 32
         %15 = OpTypePointer Function %14
         %17 = OpConstant %14 1.23000002
         %19 = OpConstant %14 1.11000001
         ... 
         %36 = OpConstant %26 5
         %37 = OpTypeArray %14 %36
         %38 = OpTypePointer Function %37
         %40 = OpConstant %14 5.5
         %41 = OpConstant %14 4.4000001
         %42 = OpConstant %14 3.29999995
         %43 = OpConstant %14 2.20000005
         %44 = OpConstant %14 1.10000002
```
In this case, ```  uint32_t maybe_float_id = fuzzerutil::MaybeGetZeroConstant(
      ir_context, transformation_context, 14, false);``` should have a value of either 17 or 19, since the type id of 14 represents a 32 bit float type, but in this case the assertion failed. Down below I also have an 5-element-array of floats. I first thought that it might return one of those ids (albeit unlikely),  but still it fails.
 I then ran 
```
  ASSERT_TRUE(std::find(float_ids.begin(), float_ids.end(), maybe_float_id)
    != float_ids.end()); 
```
and it turns out it cannot find a id with float type at all.

File: test/fuzz/fuzzerutil_test.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (afd) [reply test/fuzz/fuzzerutil_test.cpp:476]: Ah, makes sense!

For floats and negative values, the words should contain the bit pattern for the value.

You can achieve this via memcpy.

Check out line 875 of transformation_replace_constant_with_uniform_test.cpp, which does this:

```
  uint32_t float_data[1];
  float temp = 3.0;
  memcpy(&float_data[0], &temp, sizeof(float));
```

This causes float_data[0] to end up containing the bit pattern for the floating point number 3.0 (which is 0x40400000 according to [this converter](https://gregstoll.com/~gregstoll/floattohex/)).

Similarly, for a negative number you can memcpy from int32_t to uint32_t.

Note that in the above it would also be fine to do:

```
  uint32_t float_data;
  float temp = 3.0;
  memcpy(&float_data, &temp, sizeof(float));
```

The reason I used a 1 element array is for uniformity, as in other tests, for 64-bit floats, I use a 2-element array.

File: test/fuzz/fuzzerutil_test.cpp (hunk 3 above)
Comment (afd) [reply test/fuzz/fuzzerutil_test.cpp:1286]: In `fuzzerutil`, the string is just a string literal that is being anded with the assertion condition. It can be any string literal, and the point is that if the assertion fails it will appear in the assertion failure message.

In your `ASSERT_DEATH` you're saying: "This assertion should fail, and the failure string should match this regex". Now, regexes are powerful and let you write things like:

[a-z-A-Z]+

to mean: "one or more alphabetical characters".

But here, "[", "]" and "+" have special meanings. If you wanted to say "one or more characters from the set { "[", "]" } then you couldn't write:

[[]]+

Instead you'd have to write:

[\[\]]+

i.e., you'd have to escape "[" and "]" each time you want the regex to actually match that character.

This is why in death tests if you want to match "[" you have to write "\[", except that because "\" has a special meaning in C++ you have to escape *it*, hence "\\[".

Does that clear things up?

File: test/fuzz/fuzzerutil_test.cpp (hunk 4 above)
Comment (afd) [reply test/fuzz/fuzzerutil_test.cpp:1496]: In your above code, neither %17 nor %19 represent *zeros* of type float:

```
         %17 = OpConstant %14 1.23000002
```

This is the value 1.23000002 as a float.

```
         %19 = OpConstant %14 1.11000001
```

This is the value 1.11000001 as a float.

It is thus correct that `MaybeGetZeroConstant` doesn't return either of them.

A zero float constant would look like:

```
         %res = OpConstant %14 0
```

i.e., the value would be 0.

`MaybeGetZeroConstant` is implemented for floats. In the code you quoted:

```
    case SpvOpTypeFloat:
    case SpvOpTypeInt: {
      const auto width = type_inst->GetSingleWordInOperand(0);
      std::vector<uint32_t> words = {0};
      if (width > 32) {
        words.push_back(0);
      }

      return MaybeGetScalarConstant(ir_context, transformation_context, words,
                                    scalar_or_composite_type_id, is_irrelevant);
    }
```

the cases for `SpvOpTypeFloat` and `SpvOpTypeInt` lead to the same code being executed: if the opcode is `SpvOpTypeFloat` or `SpvOpTypeInt` then the code below these case lables is executed - that's because there is no `break` to stop code that takes the `SpvOpTypeFloat` case "falling through" to the `SpvOpTypeInt` case. So both cases are handled the same way.

Does that help clear things up?

File: test/fuzz/fuzzerutil_test.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (Vasniktel) [reply test/fuzz/fuzzerutil_test.cpp]: There is also a function for this: `fuzzerutil::FloatToWord`.

[the remark this answers is not in the document]
Comment (afd) [reply test/fuzz/fuzzerutil_test.cpp]: Cool - @sliu-UIUC please use that instead.

[the remark this answers is not in the document]
Comment (sliu-UIUC) [reply test/fuzz/fuzzerutil_test.cpp]: Resolved.

File: test/fuzz/fuzzerutil_test.cpp (hunk 3 above)
Comment (sliu-UIUC) [reply test/fuzz/fuzzerutil_test.cpp]: Yes! And done.

File: test/fuzz/fuzzerutil_test.cpp (hunk 4 above)
Comment (sliu-UIUC) [reply test/fuzz/fuzzerutil_test.cpp:1501]: Got it. Resolved.

File: test/fuzz/fuzzerutil_test.cpp [hunk 6]
```diff
@@ -0,0 +1,1519 @@
...
+                                               kConsoleMessageConsumer));
+  TransformationContext transformation_context(
+      MakeUnique<FactManager>(context.get()), validator_options);
+
+  opt::IRContext* ir_context = context.get();
+  //  |scalar_type_id| must be a result id of a scalar type (i.e. int, float or
+  //  bool).
+  // Returns 0 if no such
+  // instruction is present in the module. The returned id either participates
+  // in IdIsIrrelevant fact or not, depending on the |is_irrelevant| parameter.
```
[the remark this answers is not in the document]
Comment (sliu-UIUC) [reply test/fuzz/fuzzerutil_test.cpp:1086]: Finished.

File: test/fuzz/fuzzerutil_test.cpp [hunk 7]
```diff
@@ -0,0 +1,1524 @@
...
+  spvtools::ValidatorOptions validator_options;
+  ASSERT_TRUE(fuzzerutil::IsValidAndWellFormed(context.get(), validator_options,
+                                               kConsoleMessageConsumer));
+  TransformationContext transformation_context(
+      MakeUnique<FactManager>(context.get()), validator_options);
+
+  opt::IRContext* ir_context = context.get();
+  // A 32 bit float constant with width 14 exists and the id is 17.
+//   ASSERT_EQ(17, fuzzerutil::MaybeGetFloatConstant(ir_context,
+//   transformation_context, std::vector<uint32_t> {14}, 32, false));
```
Comment (Vasniktel) [inline test/fuzz/fuzzerutil_test.cpp:476]: The vector argument is supposed to specify the bit pattern for the constant. Floating-point values use different memory representations compared to integers (if you are interested in learning more, google `mantissa and exponent`). Thus, we first need to copy the bit layout from the floating-point number to the vector element. `fuzzerutil::FloatToWord` does just that.

The working assertion (hopefully :) would look like
```cpp
ASSERT_EQ(17, fuzzerutil::MaybeGetFloatConstant(ir_context,
     transformation_context, {fuzzerutil::FloatToWord(1.23000002f)}, 32, false));
```
Note that we are using the same floating-point value as in the `%17 = OpConstant %14 1.23000002`.

File: test/fuzz/fuzzerutil_test.cpp [hunk 8]
```diff
@@ -0,0 +1,1524 @@
...
+  spvtools::ValidatorOptions validator_options;
+  ASSERT_TRUE(fuzzerutil::IsValidAndWellFormed(context.get(), validator_options,
+                                               kConsoleMessageConsumer));
+  TransformationContext transformation_context(
+      MakeUnique<FactManager>(context.get()), validator_options);
+
+  opt::IRContext* ir_context = context.get();
+  uint32_t private_storage_class = SpvStorageClassPrivate;
+  uint32_t function_storage_class = SpvStorageClassFunction;
+  uint32_t input_storage_class = SpvStorageClassInput;
```
Comment (Vasniktel) [inline test/fuzz/fuzzerutil_test.cpp:955]: You can just use the `auto` type for these variables and get rid of all the `static_cast` below.

File: test/fuzz/fuzzerutil_test.cpp [hunk 9]
```diff
@@ -0,0 +1,1524 @@
...
+                                               kConsoleMessageConsumer));
+  TransformationContext transformation_context(
+      MakeUnique<FactManager>(context.get()), validator_options);
+
+  opt::IRContext* ir_context = context.get();
+  //  |scalar_type_id| must be a result id of a scalar type (i.e. int, float or
+  //  bool).
+  // Returns 0 if no such
+  // instruction is present in the module. The returned id either participates
+  // in IdIsIrrelevant fact or not, depending on the |is_irrelevant| parameter.
```
Comment (Vasniktel) [inline test/fuzz/fuzzerutil_test.cpp:1086]: I'm not sure what happened but there's no test here :).

**Review by Vasniktel — CHANGES_REQUESTED** [review]

File: test/fuzz/fuzzerutil_test.cpp [hunk 10]
```diff
@@ -0,0 +1,1524 @@
...
+                                               kConsoleMessageConsumer));
+  TransformationContext transformation_context(
+      MakeUnique<FactManager>(context.get()), validator_options);
+
+  opt::IRContext* ir_context = context.get();
+  // A float type with width = 32 and result id of 7 exists.
+  ASSERT_EQ(7, fuzzerutil::MaybeGetFloatType(ir_context, 32));
+
+  // A float int type with width = 32 exists, but the id should be 7.
+  ASSERT_FALSE(fuzzerutil::MaybeGetFloatType(ir_context, 32) == 5);
```
[the remark this answers is not in the document]
Comment (sliu-UIUC) [reply test/fuzz/fuzzerutil_test.cpp:564]: Got it. Thanks!

File: test/fuzz/fuzzerutil_test.cpp (hunk 9 above)
Comment (sliu-UIUC) [reply test/fuzz/fuzzerutil_test.cpp:1086]: Oh I think you are looking at the previous version of my PR, i.e., I didn't push the new changes to this one but the one for TransformationAccessChain. Sorry about that! I'll address some of your other comments and update this one now.

File: test/fuzz/fuzzerutil_test.cpp (hunk 8 above)
Comment (sliu-UIUC) [reply test/fuzz/fuzzerutil_test.cpp:955]: Done.

File: test/fuzz/fuzzerutil_test.cpp [hunk 11]
```diff
@@ -0,0 +1,1524 @@
...
+
+  const auto env = SPV_ENV_UNIVERSAL_1_4;
+  const auto consumer = nullptr;
+  const std::unique_ptr<opt::IRContext> context =
+      BuildModule(env, consumer, shader, kFuzzAssembleOption);
+  spvtools::ValidatorOptions validator_options;
+  ASSERT_TRUE(fuzzerutil::IsValidAndWellFormed(context.get(), validator_options,
+                                               kConsoleMessageConsumer));
+  TransformationContext transformation_context(
+      MakeUnique<FactManager>(context.get()), validator_options);
```
[the remark this answers is not in the document]
Comment (afd) [reply test/fuzz/fuzzerutil_test.cpp]: @sliu-UIUC Did you see this comment from @Vasniktel?

**Review by afd — CHANGES_REQUESTED** [review]
There are test failures. Please make sure all of your new tests pass locally, with a debug build.

File: test/fuzz/fuzzerutil_test.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (sliu-UIUC) [reply test/fuzz/fuzzerutil_test.cpp]: Yes I think those are addressed.

**Comment by sliu-UIUC** [comment]
> There are test failures. Please make sure all of your new tests pass locally, with a debug build.

Hi all my tests have passed but I do get an error "error: line 0: the number of transformations to be replayed must not exceed the size of the transformation sequence".  Do you have some insight on how to fix this?

**Review by afd — APPROVED** [review]
Thank you for this contribution!
